contracts/middle-layer/resource-mirror/AdditionalBucketHub.sol

```diff
@@ -72,49 +72,72 @@ contract AdditionalBucketHub is BucketStorage, GnfdAccessControl {
+     * It will be reset to the `msg.sender` all the time. And make sure the `refundAddress` is payable.
+     */
+    function createBucket(
+        CreateBucketSynPackage memory synPkg,
+        uint256 callbackGasLimit,
+        ExtraData memory extraData
+    ) external payable returns (bool) {
+        (uint8 _channelId, bytes memory _msgBytes, uint256 _relayFee, uint256 _ackRelayFee, ) = encodeCreateBucket(
+            msg.sender,
+            synPkg,
+            callbackGasLimit,
+            extraData
+        );
+
+        ICrossChain(CROSS_CHAIN).sendSynPackage(_channelId, _msgBytes, _relayFee, _ackRelayFee);
+        return true;
+    }
+
+    function encodeCreateBucket(
+        address sender,
+        CreateBucketSynPackage memory synPkg
+    ) public payable returns (uint8, bytes memory, uint256, uint256, address) {
         // check relay fee
         (uint256 relayFee, uint256 minAckRelayFee) = ICrossChain(CROSS_CHAIN).getRelayFees();
         require(msg.value >= relayFee + minAckRelayFee, "not enough fee");
         uint256 _ackRelayFee = msg.value - relayFee;
 
         // check authorization
         address owner = synPkg.creator;
-        if (msg.sender != owner) {
-            require(hasRole(ROLE_CREATE, owner, msg.sender), "no permission to create");
+        if (sender != owner) {
+            require(hasRole(ROLE_CREATE, owner, sender), "no permission to create");
         }
 
         // make sure the extra data is as expected
         synPkg.extraData = "";
 
-        ICrossChain(CROSS_CHAIN).sendSynPackage(
-            BUCKET_CHANNEL_ID,
-            abi.encodePacked(TYPE_CREATE, abi.encode(synPkg)),
-            relayFee,
-            _ackRelayFee
-        );
-
         // transfer all the fee to tokenHub
         (bool success, ) = TOKEN_HUB.call{ value: address(this).balance }("");
         require(success, "transfer to tokenHub failed");
 
         emit CreateSubmitted(owner, msg.sender, synPkg.name);
-        return true;
+
+        return (BUCKET_CHANNEL_ID, abi.encodePacked(TYPE_CREATE, abi.encode(synPkg)), relayFee, _ackRelayFee, sender);
     }
 
-    /**
-     * @dev create a bucket and send cross-chain request from BSC to GNFD.
-     * Callback function will be called when the request is processed.
-     *
-     * @param synPkg Package containing information of the bucket to be created
-     * @param callbackGasLimit The gas limit for callback function
-     * @param extraData Extra data for callback function. The `appAddress` in `extraData` will be ignored.
-     * It will be reset to the `msg.sender` all the time. And make sure the `refundAddress` is payable.
-     */
-    function createBucket(
+    function encodeCreateBucket(
+        address sender,
         CreateBucketSynPackage memory synPkg,
         uint256 callbackGasLimit,
         ExtraData memory extraData
-    ) external payable returns (bool) {
+    ) public payable returns (uint8, bytes memory, uint256, uint256, address) {
```

Comment:
this function should not be public.

Reply:
how can a channel id be an outputs of an `encode` function.

To me, the outputs of an `encode` function must be related with the inputs.

Reply:
modified